File: backend/nodejs/jsonDb.js
const fs = require('fs');
const path = require('path');
const StreamArray = require('stream-json/streamers/StreamArray');

const filePath = path.join(__dirname, '../foodNutrientDatabase_trimmed.json');

function safeGet(obj, keys) {
  for (const k of keys) {
    if (obj && Object.prototype.hasOwnProperty.call(obj, k)) return obj[k];
  }
  return undefined;
}

function mapNutrients(item) {
  const out = {
    calories: 0, protein: 0, carbs: 0, fat: 0, fiber: 0, sugar: 0,
    calcium: 0, iron: 0, magnesium: 0, phosphorus: 0, potassium: 0, sodium: 0, zinc: 0,
    vitaminA: 0, vitaminC: 0, vitaminD: 0, vitaminE: 0, vitaminK: 0,
    vitaminB6: 0, vitaminB12: 0, folate: 0, niacin: 0
  };

  const nutrients = safeGet(item, ['foodNutrients', 'nutrients', 'nutrient', 'nutrient_values', 'nutrient_values_list']);

  if (Array.isArray(nutrients)) {
    for (const n of nutrients) {
      // USDA FDC: nutrient code is in n.nutrient.number (string like "208", "203", etc.)
      let code = n.nutrient?.number || n.nutrient_code || n.code || n.id || n.nutrientId || n.nutrient;
      if (code) code = String(code); 
      
      const value = Number(n.amount ?? n.value ?? n.nutrient_value ?? n.quantity ?? 0) || 0;
      const nutName = (n.nutrient?.name || n.name || n.nutrient_name || '').toString().toLowerCase();

      // Macronutrients
      if (code === '208' || nutName.includes('energy') && nutName.includes('kcal')) out.calories = value;
      if (code === '203' || nutName.includes('protein')) out.protein = value;
      if (code === '205' || nutName.includes('carbohydrate')) out.carbs = value;
      if (code === '204' || nutName.includes('lipid') || nutName.includes('fat')) out.fat = value;
      if (code === '291' || nutName.includes('fiber')) out.fiber = value;
      if (code === '269' || nutName.includes('sugar')) out.sugar = value;
      
      // Minerals
      if (code === '301' || nutName.includes('calcium')) out.calcium = value;
      if (code === '303' || nutName.includes('iron')) out.iron = value;
      if (code === '304' || nutName.includes('magnesium')) out.magnesium = value;
      if (code === '305' || nutName.includes('phosphorus')) out.phosphorus = value;
      if (code === '306' || nutName.includes('potassium')) out.potassium = value;
      if (code === '307' || nutName.includes('sodium')) out.sodium = value;
      if (code === '309' || nutName.includes('zinc')) out.zinc = value;
      
      // Vitamins
      if (code === '320' || nutName.includes('vitamin a') && nutName.includes('rae')) out.vitaminA = value;
      if (code === '401' || nutName.includes('vitamin c')) out.vitaminC = value;
      if (code === '328' || nutName.includes('vitamin d')) out.vitaminD = value;
      if (code === '323' || nutName.includes('vitamin e') && nutName.includes('tocopherol')) out.vitaminE = value;
      if (code === '430' || nutName.includes('vitamin k')) out.vitaminK = value;
      if (code === '415' || nutName.includes('vitamin b-6')) out.vitaminB6 = value;
      if (code === '418' || nutName.includes('vitamin b-12')) out.vitaminB12 = value;
      if (code === '417' || nutName.includes('folate')) out.folate = value;
      if (code === '406' || nutName.includes('niacin')) out.niacin = value;
    }
  } else if (nutrients && typeof nutrients === 'object') {
    const pick = (keys) => {
      for (const k of keys) if (nutrients[k] != null) return Number(nutrients[k]) || 0;
      return 0;
    };
    out.calories = pick(['calories', 'energy', 'kcal']);
    out.protein = pick(['protein']);
    out.carbs = pick(['carbs', 'carbohydrate', 'carbohydrates']);
    out.fat = pick(['fat', 'fats']);
    out.fiber = pick(['fiber', 'fibre']);
    out.sugar = pick(['sugar']);
    out.calcium = pick(['calcium']);
    out.iron = pick(['iron']);
    out.magnesium = pick(['magnesium']);
    out.phosphorus = pick(['phosphorus']);
    out.potassium = pick(['potassium']);
    out.sodium = pick(['sodium']);
    out.zinc = pick(['zinc']);
    out.vitaminA = pick(['vitamin_a', 'vitaminA']);
    out.vitaminC = pick(['vitamin_c', 'vitaminC']);
    out.vitaminD = pick(['vitamin_d', 'vitaminD']);
    out.vitaminE = pick(['vitamin_e', 'vitaminE']);
    out.vitaminK = pick(['vitamin_k', 'vitaminK']);
    out.vitaminB6 = pick(['vitamin_b6', 'vitaminB6']);
    out.vitaminB12 = pick(['vitamin_b12', 'vitaminB12']);
    out.folate = pick(['folate']);
    out.niacin = pick(['niacin']);
  }

  return out;
}

// Standard measurement options presented to users for every food item.
// Each option includes an approximate `gramsPerUnit` when sensible. Frontend
// can use `gramsPerUnit` to convert the entered numeric amount into grams;
// options without `gramsPerUnit` should be treated as "eyeball" units.
const STANDARD_SERVINGS = [
  { label: 'g', type: 'mass', unit: 'g', gramsPerUnit: 1 },
  { label: 'oz', type: 'mass', unit: 'oz', gramsPerUnit: 28.35 },
  { label: 'cup', type: 'volume', unit: 'cup', gramsPerUnit: 240 },
  { label: 'tbsp', type: 'volume', unit: 'tbsp', gramsPerUnit: 15 },
  { label: 'tsp', type: 'volume', unit: 'tsp', gramsPerUnit: 5 },
  { label: 'bowl', type: 'eyeball', unit: 'bowl', gramsPerUnit: 300 },
  { label: 'piece', type: 'eyeball', unit: 'piece', gramsPerUnit: null }
];

function mapItem(item) {
  const id = safeGet(item, ['fdcId', 'ndbNumber', 'cn_code', 'id', 'code', 'food_id']) || null;
  const name = safeGet(item, ['description', 'descriptor', 'name', 'desc', 'food_name']) || 'Unknown';

  const servingOptions = STANDARD_SERVINGS.map(o => ({ ...o }));
  const defaultServing = { label: 'g', type: 'mass', unit: 'g', gramsPerUnit: 1 };
  const unit = `100 ${defaultServing.label}`;

  const nut = mapNutrients(item);

  return {
    id,
    name,
    unit,
    servingOptions,
    calories: nut.calories,
    protein: nut.protein,
    carbs: nut.carbs,
    fat: nut.fat,
    fiber: nut.fiber,
    sugar: nut.sugar,
    calcium: nut.calcium,
    iron: nut.iron,
    magnesium: nut.magnesium,
    phosphorus: nut.phosphorus,
    potassium: nut.potassium,
    sodium: nut.sodium,
    zinc: nut.zinc,
    vitaminA: nut.vitaminA,  
    vitaminC: nut.vitaminC,      
    vitaminD: nut.vitaminD,      
    vitaminE: nut.vitaminE,      
    vitaminK: nut.vitaminK,      
    vitaminB6: nut.vitaminB6,    
    vitaminB12: nut.vitaminB12,  
    folate: nut.folate,          
    niacin: nut.niacin           
  };
}

function streamFind(limit = 100) {
  return new Promise((resolve, reject) => {
    const results = [];
    const stream = fs.createReadStream(filePath).pipe(StreamArray.withParser());

    stream.on('data', ({key, value}) => {
      results.push(mapItem(value));
      if (results.length >= limit) {
        stream.destroy();
        resolve(results);
      }
    });

    stream.on('end', () => resolve(results));
    stream.on('error', (err) => {
      if (err && /top[- ]level object should be an array/i.test(err.message)) {
        try {
          const data = fs.readFileSync(filePath, 'utf8');
          const parsed = JSON.parse(data);
          let arr = null;
          if (Array.isArray(parsed)) arr = parsed;
          else if (parsed && typeof parsed === 'object') {
            for (const k of Object.keys(parsed)) {
              if (Array.isArray(parsed[k])) {
                arr = parsed[k];
                break;
              }
            }
          }

          if (!arr) return reject(new Error('No array found inside JSON top-level object'));

          const sliced = arr.slice(0, limit).map(mapItem);
          return resolve(sliced);
        } catch (e) {
          return reject(e);
        }
      }
      return reject(err);
    });
  });
}

function streamSearch(query, limit = 50) {
  return new Promise((resolve, reject) => {
    const normalize = (s) => (s || '').toString().toLowerCase().replace(/[^a-z0-9]+/g, ' ').trim();
    const q = normalize(query);
    const tokens = q.split(/\s+/).filter(Boolean);
    const results = [];
    const stream = fs.createReadStream(filePath).pipe(StreamArray.withParser());

    stream.on('data', ({key, value}) => {
      const nameRaw = (safeGet(value, ['description','descriptor','name','desc','food_name']) || '').toString();
      const name = normalize(nameRaw);
      const ok = tokens.length === 0 ? false : tokens.every(t => name.includes(t));
      if (ok) results.push(mapItem(value));
      if (results.length >= limit) {
        stream.destroy();
        resolve(results);
      }
    });

    stream.on('end', () => {
      try {
        const data = fs.readFileSync(filePath, 'utf8');
        const parsed = JSON.parse(data);
        let arr = null;
        if (Array.isArray(parsed)) arr = parsed;
        else if (parsed && typeof parsed === 'object') {
          for (const k of Object.keys(parsed)) {
            if (Array.isArray(parsed[k])) {
              arr = parsed[k];
              break;
            }
          }
        }

        if (!arr) {
          return resolve(results);
        }

        for (const value of arr) {
          const name = normalize(safeGet(value, ['description','descriptor','name','desc','food_name']) || '');
          if (tokens.length > 0 && tokens.every(t => name.includes(t))) {
            results.push(mapItem(value));
            if (results.length >= limit) break;
          }
        }
        return resolve(results);
      } catch (e) {
        return resolve(results);
      }
    });

    stream.on('error', (err) => {
      if (err && /top[- ]level object should be an array/i.test(err.message)) {
        try {
          const data = fs.readFileSync(filePath, 'utf8');
          const parsed = JSON.parse(data);
          let arr = null;
          if (Array.isArray(parsed)) arr = parsed;
          else if (parsed && typeof parsed === 'object') {
            for (const k of Object.keys(parsed)) {
              if (Array.isArray(parsed[k])) {
                arr = parsed[k];
                break;
              }
            }
          }

          if (!arr) return reject(new Error('No array found inside JSON top-level object'));

          for (const value of arr) {
            const name = normalize(safeGet(value, ['description','descriptor','name','desc','food_name']) || '');
            if (tokens.length > 0 && tokens.every(t => name.includes(t))) {
              results.push(mapItem(value));
              if (results.length >= limit) break;
            }
          }

          return resolve(results);
        } catch (e) {
          return reject(e);
        }
      }
      return reject(err);
    });
  });
}

module.exports = { streamFind, streamSearch, mapItem };

File: backend/json_db.py
#!/usr/bin/env python3
"""
JSON Database module for ATE Nutrition Tracking App
Equivalent to jsonDb.js - handles USDA FDC JSON format

database source: https://www.fns.usda.gov/usda-fis/usda-foods-database
"""

import json
from models.food import Food, ServingOption

# USDA FDC Nutrient Code Mapping
NUTRIENT_CODE_MAP = {
    208: 'calories',
    203: 'protein',
    205: 'carbs',
    204: 'fat',
    291: 'fiber',
    269: 'sugar',
    301: 'calcium',
    303: 'iron',
    304: 'magnesium',
    305: 'phosphorus',
    306: 'potassium',
    307: 'sodium',
    309: 'zinc',
    320: 'vitaminA',
    401: 'vitaminC',
    328: 'vitaminD',
    323: 'vitaminE',
    430: 'vitaminK',
    415: 'vitaminB6',
    418: 'vitaminB12',
    417: 'folate',
    406: 'niacin'
}


class JsonDatabase:
    """JSON-based food database using USDA FDC format"""
    
    def __init__(self, json_file_path):
        """Initialize database from JSON file"""
        self.foods = []
        self.load_from_file(json_file_path)
    
    def load_from_file(self, json_file_path):
        """Load foods from JSON file"""
        try:
            with open(json_file_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
                
                # Handle different JSON structures
                if isinstance(data, list):
                    self.foods = data
                elif isinstance(data, dict) and 'foods' in data:
                    self.foods = data['foods']
                elif isinstance(data, dict) and 'FoundationFoods' in data:
                    self.foods = data['FoundationFoods']
                else:
                    self.foods = []
                
                print(f"Loaded {len(self.foods)} foods from JSON database")
        except FileNotFoundError:
            print(f"JSON file not found: {json_file_path}")
            self.foods = []
        except json.JSONDecodeError as e:
            print(f"Error parsing JSON: {e}")
            self.foods = []
        except Exception as e:
            print(f"Error loading JSON database: {e}")
            self.foods = []
    
    def _extract_nutrient(self, food, nutrient_code):
        """Extract nutrient value from food object"""
        nutrients = food.get('foodNutrients', [])
        for nutrient in nutrients:
            # Handle different field names
            nutrient_id = nutrient.get('nutrientId') or nutrient.get('nutrient', {}).get('id')
            # Also handle 'number' field (as string)
            if not nutrient_id:
                nutrient_number = nutrient.get('nutrient', {}).get('number')
                if nutrient_number:
                    try:
                        nutrient_id = int(nutrient_number)
                    except (ValueError, TypeError):
                        pass
            
            if nutrient_id == nutrient_code:
                amount = nutrient.get('amount') or nutrient.get('value', 0)
                return float(amount) if amount is not None else 0.0
        return 0.0
    
    def _normalize_food(self, food):
        """Convert USDA food format to app format with Food model"""
        # Extract basic info
        food_id = food.get('fdcId') or food.get('fdc_id') or food.get('id', 0)
        description = food.get('description') or food.get('name', 'Unknown Food')
        
        # Build serving options using ServingOption model
        serving_options = [
            ServingOption('grams', 'g', 1, 1),
            ServingOption('oz', 'oz', 28.35, 28.35),
            ServingOption('lb', 'lb', 453.59, 453.59),
            ServingOption('serving (100g)', 'serving', 100, 100),
        ]
        
        # Add food-specific serving options from foodPortions
        portions = food.get('foodPortions', [])
        for portion in portions:
            label = portion.get('label') or portion.get('portionDescription') or portion.get('modifier') or 'serving'
            amount = portion.get('amount', 1.0)
            gram_weight = portion.get('gramWeight', 100.0)
            
            final_label = f"{amount} {label}" if amount != 1.0 else label
            grams_per_unit = gram_weight / amount if amount > 0 else gram_weight
            serving_options.append(ServingOption(final_label, label, grams_per_unit, gram_weight))
        
        # Extract all nutrients using the mapping
        nutrients = {}
        for code, name in NUTRIENT_CODE_MAP.items():
            nutrients[name] = self._extract_nutrient(food, code)
        
        # Create Food model instance
        food_model = Food(
            food_id=food_id,
            name=description,
            unit='100 g',
            calories=nutrients.get('calories', 0.0),
            protein=nutrients.get('protein', 0.0),
            carbs=nutrients.get('carbs', 0.0),
            fat=nutrients.get('fat', 0.0),
            fiber=nutrients.get('fiber', 0.0),
            sugar=nutrients.get('sugar', 0.0),
            calcium=nutrients.get('calcium', 0.0),
            iron=nutrients.get('iron', 0.0),
            magnesium=nutrients.get('magnesium', 0.0),
            phosphorus=nutrients.get('phosphorus', 0.0),
            potassium=nutrients.get('potassium', 0.0),
            sodium=nutrients.get('sodium', 0.0),
            zinc=nutrients.get('zinc', 0.0),
            vitamin_a=nutrients.get('vitaminA', 0.0),
            vitamin_c=nutrients.get('vitaminC', 0.0),
            vitamin_d=nutrients.get('vitaminD', 0.0),
            vitamin_e=nutrients.get('vitaminE', 0.0),
            vitamin_k=nutrients.get('vitaminK', 0.0),
            vitamin_b6=nutrients.get('vitaminB6', 0.0),
            vitamin_b12=nutrients.get('vitaminB12', 0.0),
            folate=nutrients.get('folate', 0.0),
            niacin=nutrients.get('niacin', 0.0),
            serving_options=serving_options
        )
        
        # Return as dictionary for backward compatibility with existing code
        return food_model.to_dict()
    
    def search(self, query, limit=20):
        """Search foods by text query - matches all words in any order"""
        if not query:
            return []
        
        query_lower = query.lower().strip()
        # Split query into words for flexible matching
        query_words = query_lower.split()
        results = []
        
        for food in self.foods:
            description = food.get('description', '') or food.get('name', '')
            description_lower = description.lower()
            
            # Check if all query words are present in the description
            if all(word in description_lower for word in query_words):
                results.append(self._normalize_food(food))
                if len(results) >= limit:
                    break
        
        return results
    
    def get_by_id(self, food_id):
        """Get food by ID"""
        for food in self.foods:
            fdc_id = food.get('fdcId') or food.get('fdc_id') or food.get('id')
            if fdc_id == food_id:
                return self._normalize_food(food)
        return None
    
    def get_all(self, limit=50):
        """Get all foods (limited)"""
        results = []
        for food in self.foods[:limit]:
            results.append(self._normalize_food(food))
        return results


# Standalone test function
def test_json_db():
    """Test the JSON database"""
    import os
    
    # Try to find JSON file
    json_file = os.path.join(os.path.dirname(__file__), 'usda_foods.json')
    
    if not os.path.exists(json_file):
        print(f"JSON file not found: {json_file}")
        print("Please create a usda_foods.json file with USDA FDC data")
        return
    
    # Initialize database
    db = JsonDatabase(json_file)
    
    # Test search
    print("\nTesting search for 'spinach':")
    results = db.search('spinach', limit=3)
    for food in results:
        print(f"  - {food['name']}")
        print(f"    Calories: {food['calories']}, Protein: {food['protein']}g")
        print(f"    Iron: {food['iron']}mg, Potassium: {food['potassium']}mg")
    
    print("\nTesting search for 'banana':")
    results = db.search('banana', limit=3)
    for food in results:
        print(f"  - {food['name']}")
        print(f"    Calories: {food['calories']}, Carbs: {food['carbs']}g")
        print(f"    Potassium: {food['potassium']}mg")


if __name__ == '__main__':
    test_json_db()
